### ISSUE TITLE
Firefox nightly install error

### ISSUE THREAD

**Issue by adeptnerd** [Tobi823/ffupdater#351]
Something similar happened previously like a month ago and it was fixed very quickly. 

Now I don't know if I did something wrong, I was trying to install Firefox but my internet abruptly dropped and it caused this error, I suppose. And now I keep getting this error. I tried clearing cache, reinstalled the app, but I still get the same error. 

Stacktrace:
```
java.lang.Exception: Can't find latest update for FIREFOX_NIGHTLY.
	at de.marmaro.krt.ffupdater.storage.MetadataCache.getCachedOrFetchIfOutdated(MetadataCache.kt:33)
	at de.marmaro.krt.ffupdater.storage.MetadataCache$getCachedOrFetchIfOutdated$1.invokeSuspend(MetadataCache.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:226)
	at android.os.Looper.loop(Looper.java:313)
	at android.app.ActivityThread.main(ActivityThread.java:8741)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
Caused by: java.lang.Exception: Can't find latest update for FirefoxNightly.
	at de.marmaro.krt.ffupdater.app.impl.AppBase.findAppUpdateStatus(AppBase.kt:96)
	at de.marmaro.krt.ffupdater.app.impl.AppBase$findAppUpdateStatus$1.invokeSuspend(AppBase.kt:0)
	... 10 more
Caused by: java.lang.NullPointerException
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer.updateCheck(MozillaCiJsonConsumer.kt:35)
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer$updateCheck$1.invokeSuspend(MozillaCiJsonConsumer.kt:0)
	... 10 more
```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.6 (132) release |
| Device | SM-M536B (m53xnnxx, m53x, m53x) |
| Manufacturer | samsung (samsung) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 13 (SDK: 33) |
| OS | SWDK6923, dpi, release-keys, 1673856816000 |

**Comment by U-Gautam** [comment]
same problem as yours. Thank you for posting

**Comment by ILeMeNiiz** [comment]
Same here

**Comment by adeptnerd** [comment]
Stacktrace:
```
de.marmaro.krt.ffupdater.background.BackgroundException: The background job fails due to an unrecoverable exception.
	at de.marmaro.krt.ffupdater.BackgroundJob.handleDoWorkException(BackgroundJob.kt:91)
	at de.marmaro.krt.ffupdater.BackgroundJob.access$handleDoWorkException(BackgroundJob.kt:46)
	at de.marmaro.krt.ffupdater.BackgroundJob$doWork$2.invokeSuspend(BackgroundJob.kt:82)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:749)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
Caused by: java.lang.Exception: Can't find latest update for FIREFOX_NIGHTLY.
	at de.marmaro.krt.ffupdater.storage.MetadataCache.getCachedOrFetchIfOutdated(MetadataCache.kt:33)
	at de.marmaro.krt.ffupdater.storage.MetadataCache$getCachedOrFetchIfOutdated$1.invokeSuspend(MetadataCache.kt:0)
	... 6 more
Caused by: java.lang.Exception: Can't find latest update for FirefoxNightly.
	at de.marmaro.krt.ffupdater.app.impl.AppBase.findAppUpdateStatus(AppBase.kt:96)
	at de.marmaro.krt.ffupdater.app.impl.AppBase$findAppUpdateStatus$1.invokeSuspend(AppBase.kt:0)
	... 6 more
Caused by: java.lang.NullPointerException
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer.updateCheck(MozillaCiJsonConsumer.kt:35)
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer$updateCheck$1.invokeSuspend(MozillaCiJsonConsumer.kt:0)
	... 6 more
```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.6 (132) release |
| Device | SM-M536B (m53xnnxx, m53x, m53x) |
| Manufacturer | samsung (samsung) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 13 (SDK: 33) |
| OS | SWDK6923, dpi, release-keys, 1673856816000 |         

I just got this issue and the app asked me to report it. I don't know what's wrong. Hopefully you will figure it out, lol

**Comment by ILeMeNiiz** [comment]
Stacktrace:
```
java.lang.Exception: Can't find latest update for FIREFOX_NIGHTLY.
	at de.marmaro.krt.ffupdater.storage.MetadataCache.getCachedOrFetchIfOutdated(MetadataCache.kt:33)
	at de.marmaro.krt.ffupdater.storage.MetadataCache$getCachedOrFetchIfOutdated$1.invokeSuspend(MetadataCache.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7872)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@f40b600, Dispatchers.Main]
Caused by: java.lang.Exception: Can't find latest update for FirefoxNightly.
	at de.marmaro.krt.ffupdater.app.impl.AppBase.findAppUpdateStatus(AppBase.kt:96)
	at de.marmaro.krt.ffupdater.app.impl.AppBase$findAppUpdateStatus$1.invokeSuspend(AppBase.kt:0)
	... 10 more
Caused by: java.lang.NullPointerException
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer.updateCheck(MozillaCiJsonConsumer.kt:35)
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer$updateCheck$1.invokeSuspend(MozillaCiJsonConsumer.kt:0)
	... 10 more
Timestamp: 2023-02-26T17:55:51.685

```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.6 (132) release |
| Device | Pixel 6 Pro (raven, raven, raven) |
| Manufacturer | google (Google) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 13 (SDK: 33) |
| OS | abfarm-release-2004-0100, android-build, release-keys, 1673310463000 |

**Comment by L31FY** [comment]
Issue #354 seems linked or identical

**Comment by adeptnerd** [comment]
Yeah, it seems identical or at least linked. And even the 77.7.7 update doesn't seem to fix the issue for me.

**Comment by APPLESOFT98** [comment]
I have the same problem while downloading Firefox Nightly and Chromium (issue [#340](https://github.com/Tobi823/ffupdater/issues/340)). With Firefox Nightly it just crash but while downloading Chromium it reaches 100% and then crash. One thing in common I see is we get this only with Android 13 devices

Stacktrace:
```
java.lang.Exception: Can't find latest update for FIREFOX_NIGHTLY.
	at de.marmaro.krt.ffupdater.storage.MetadataCache.getCachedOrFetchIfOutdated(MetadataCache.kt:33)
	at de.marmaro.krt.ffupdater.storage.MetadataCache$getCachedOrFetchIfOutdated$1.invokeSuspend(MetadataCache.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:226)
	at android.os.Looper.loop(Looper.java:313)
	at android.app.ActivityThread.main(ActivityThread.java:8757)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1067)
	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@bd633c1, Dispatchers.Main]
Caused by: java.lang.Exception: Can't find latest update for FirefoxNightly.
	at de.marmaro.krt.ffupdater.app.impl.AppBase.findAppUpdateStatus(AppBase.kt:96)
	at de.marmaro.krt.ffupdater.app.impl.AppBase$findAppUpdateStatus$1.invokeSuspend(AppBase.kt:0)
	... 10 more
Caused by: java.lang.NullPointerException
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer.updateCheck(MozillaCiJsonConsumer.kt:35)
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer$updateCheck$1.invokeSuspend(MozillaCiJsonConsumer.kt:0)
	... 10 more
Timestamp: 2023-02-27T23:55:24.680

```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.7 (133) release |
| Device | SM-G991B (o1sxeea, o1s, exynos2100) |
| Manufacturer | samsung (samsung) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 13 (SDK: 33) |
| OS | SWDK6118, dpi, release-keys, 1675073918000 |

**Comment by adeptnerd** [comment]
Yeah, I tried with an android 11 device as well. It's not related to android 13.

Stacktrace:
```
java.lang.Exception: Can't find latest update for FIREFOX_NIGHTLY.
	at de.marmaro.krt.ffupdater.storage.MetadataCache.getCachedOrFetchIfOutdated(MetadataCache.kt:33)
	at de.marmaro.krt.ffupdater.storage.MetadataCache$getCachedOrFetchIfOutdated$1.invokeSuspend(MetadataCache.kt:0)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:247)
	at android.app.ActivityThread.main(ActivityThread.java:8676)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
	Suppressed: kotlinx.coroutines.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@317dfc4, Dispatchers.Main]
Caused by: java.lang.Exception: Can't find latest update for FirefoxNightly.
	at de.marmaro.krt.ffupdater.app.impl.AppBase.findAppUpdateStatus(AppBase.kt:96)
	at de.marmaro.krt.ffupdater.app.impl.AppBase$findAppUpdateStatus$1.invokeSuspend(AppBase.kt:0)
	... 9 more
Caused by: java.lang.NullPointerException
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer.updateCheck(MozillaCiJsonConsumer.kt:35)
	at de.marmaro.krt.ffupdater.network.mozillaci.MozillaCiJsonConsumer$updateCheck$1.invokeSuspend(MozillaCiJsonConsumer.kt:0)
	... 9 more
Timestamp: 2023-02-28T23:10:28.635

```
Device information:
| Key | Value |
| --- | --- |
| FFUpdater version | 77.7.7 (133) release |
| Device | SM-T720 (gts4lvwifixx, gts4lvwifi, sdm710) |
| Manufacturer | samsung (samsung) | 
| Supported ABIs | arm64-v8a, armeabi-v7a, armeabi |
| Android version | 11 (SDK: 30) |
| OS | VPDJR1012, dpi, release-keys, 1648622431000 |

**Comment by Tobi823** [comment]
I can reproduce the error - maybe some invalid update file was cached? @adeptnerd can you delete the cache and storage of FFUpdater and try to install Firefox Nightly?

**Comment by Tobi823** [comment]
Nevermind, this is not the problem

**Comment by Tobi823** [comment]
I can only reproduce the bug with the release version of FFUpdater. I suspect that Proguard is the culprit.

**Comment by Tobi823** [comment]
Yes, it is the culprit. I need an additional Proguard rule for Gson

**Comment by Tobi823** [comment]
Should be fixed in 77.7.8
